extensions/vscode-extensiontest/src/extension.ts

```diff
@@ -0,0 +1,58 @@
+import { ExtensionContext, ReferenceProvider, ReferenceContext, TextDocument, CancellationToken, ProviderResult, Location, Position, Uri, languages } from 'vscode';
+
+export function activate(context: ExtensionContext) {
+
+	const goReferenceProvider = new GoReferenceProvider();
+	languages.registerReferenceProvider("go", goReferenceProvider);
+	context.subscriptions.push(goReferenceProvider);
+}
+
+export class GoReferenceProvider implements ReferenceProvider {
+
+	provideReferences(document: TextDocument, position: Position, context: ReferenceContext, token: CancellationToken, progress: (locations: Location[]) => void): ProviderResult<Location[]> {
+		console.log("provide references");
+		return new Promise<Location[]>((resolve, reject) => {
```

Comment:
@nicksnyder not a big deal, but just in the interest of education you can do this:

```
Promise.all(locations.reverse().reduce(...)).then(() => locations)
```

You could also change the reduce from returning `Promise<void>` to `Promise<Location[]>` like this:

```
locations.reverse.reduce((p, l, i) => p.then((locs) => resolve(locs.concat(loc)), Promise.resolve([]))
```

`Promise.all` FTW